ge/org.osate.ge/src/org/osate/ge/internal/ui/editor/DeleteFlowContributionItem.java

```diff
@@ -0,0 +1,163 @@
+
+public class DeleteFlowContributionItem extends ControlContribution {
+	private final static Image deleteImage = PlatformUI.getWorkbench().getSharedImages()
+			.getImage(ISharedImages.IMG_TOOL_DELETE);
+	private AgeDiagramEditor editor = null;
+	private Button deleteFlowBtn;
+	private HighlightableFlowInfo selectedFlow;
+
+	protected DeleteFlowContributionItem(final String id) {
+		super(id);
+	}
+
+
+	@Override
+	protected Control createControl(final Composite parent) {
+		deleteFlowBtn = new Button(parent, SWT.PUSH);
+		deleteFlowBtn.setImage(deleteImage);
+		deleteFlowBtn.setToolTipText("Delete");
+
+		updateButton();
+		deleteFlowBtn.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(final SelectionEvent e) {
+				if (editor != null && selectedFlow != null) {
+					final AadlModificationService aadlModificationService = Objects.requireNonNull(
+							(AadlModificationService) editor.getAdapter(AadlModificationService.class),
+							"Unable to retrieve modification service");
+					final NamedElement ne = getFlowToRemove();
+					// Delete
+					aadlModificationService.modify(ne, EcoreUtil::remove);
+				}
+			}
+
+			/**
+			 * Get the flow to modify/remove
+			 * @return the flow to modify/remove
+			 */
+			private NamedElement getFlowToRemove() {
+				final NamedElement selectedFlowSegment = selectedFlow.getFlowSegment();
+				final Queryable container = selectedFlow.getContainer();
+				final ComponentImplementation componentImpl = FlowContributionItemUtil
+						.getComponentImplementation(container);
+				if (selectedFlowSegment instanceof EndToEndFlow) {
+					final ComponentImplementation endToEndFlowContainer = ProxyUtil
+							.resolveOrNull(
+									selectedFlowSegment.getContainingComponentImpl(), ComponentImplementation.class,
+									componentImpl.eResource().getResourceSet());
+					return ProxyUtil.resolveOrNull(selectedFlowSegment, EndToEndFlow.class,
+							endToEndFlowContainer.eResource().getResourceSet());
+				} else if (selectedFlowSegment instanceof FlowSpecification) {
+					final FlowSpecification flowSpecification = ProxyUtil.resolveOrNull(selectedFlowSegment,
+							FlowSpecification.class, componentImpl.eResource().getResourceSet());
+					// If there are multiple flow implementations for the flow spec, choose desired one.
+					final List<FlowImplementation> flowImpls = componentImpl.getOwnedFlowImplementations().stream()
+							.filter(fi -> fi.getSpecification() == flowSpecification).collect(Collectors.toList());
+					return getFlowImplementation(flowImpls).orElseThrow(() -> new RuntimeException(
+							"Cannot find flow implementation for flow specification: "
+									+ FlowContributionItemUtil.getName(flowSpecification)));
+				} else {
+					throw new RuntimeException("Unexpected flow type: " + selectedFlow.getFlowSegment());
+				}
+			}
+
+			private Optional<FlowImplementation> getFlowImplementation(final List<FlowImplementation> flowImpls) {
```

Comment:
During testing, all flow implementations had the name "<unknown>".  All I did is create 2 flow source implementations and then tried to delete. In practice, the name doesn't make sense in this case. Multiple flow implementations will have the same name. It could be named based like "spec1 implementation #1 (<Component Implementation>)" or something. Not sure what is practical.

Reply:
Just to make sure I understand, when I tested this shows up as "unknown", which is something I broke while cleaning up and I have now fixed it.

Reply:
I have added the "#" sign and index as well.

Reply:
I'll retest during the next review.

Reply:
Okay, I'll go ahead and push my changes.